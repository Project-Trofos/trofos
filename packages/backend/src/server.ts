import express from 'express';
import cors from 'cors';
import cookieParser from 'cookie-parser';
import accountRouter from './routes/account.route';
import courseRouter from './routes/course.route';
import projectRouter from './routes/project.route';
import backlogRouter from './routes/backlog.route';
import userRouter from './routes/user.route';
import sprintRouter from './routes/sprint.route';
import feedbackRouter from './routes/feedback.route';
import roleRouter from './routes/role.route';
import githubRouter from './routes/github.route';
import settingsRouter from './routes/settings.route';

const app = express();

export const port = 3001;
export const frontendUrl = process.env.FRONTEND_BASE_URL || 'http://localhost:3000';

export const corsOptions = {
  origin: frontendUrl,
  credentials: true,
};

app.use(express.json());
app.use(cookieParser());
app.use(cors(corsOptions));

app.get('/', (req: express.Request, res: express.Response) => {
  res.send('Hello World!');
});

const router = express.Router()

router.use('/account', accountRouter);

// Routes for course
router.use('/course', courseRouter);

// Routes for project
router.use('/project', projectRouter);

// Routes for backlog
router.use('/backlog', backlogRouter);

// Routes for user
router.use('/user', userRouter);

// Routes for backlog
router.use('/sprint', sprintRouter);

// Routes for settings
router.use('/feedback', feedbackRouter);

// Routes for role
router.use('/role', roleRouter);

// Routes for github app
router.use('/github', githubRouter);

// Routes for settings
router.use('/settings', settingsRouter);

app.use('/api', router)
// For unit testing
export default app;
