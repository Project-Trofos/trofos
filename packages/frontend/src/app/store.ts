import { configureStore } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';
import trofosApiSlice, { nusmodsApiSlice } from '../api';

const store = configureStore({
  reducer: {
    [trofosApiSlice.reducerPath]: trofosApiSlice.reducer,
    [nusmodsApiSlice.reducerPath]: nusmodsApiSlice.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(trofosApiSlice.middleware).concat(nusmodsApiSlice.middleware),
});

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
