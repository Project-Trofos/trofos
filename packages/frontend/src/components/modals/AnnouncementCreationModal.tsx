import React, { useCallback } from 'react';
import { Form, message } from 'antd';
import { useCreateAnnouncementMutation } from '../../api/course';
import MultistepFormModal from './MultistepModalForm';
import { getErrorMessage } from '../../helpers/error';
import StringFormItem from '../forms/StringFormItem';

/**
 * Modal for creating announcement
 */
export default function AnnouncementCreationModal({ courseId }: { courseId: string }) {
  const [createAnnouncement] = useCreateAnnouncementMutation();

  const [form] = Form.useForm();

  const onFinish = useCallback(
    async (values: { announcementTitle?: string; announcementContent?: string }) => {
      try {
        const { announcementContent, announcementTitle } = values;

        if (!announcementTitle || !announcementContent) {
          throw new Error('Please input valid title or content!');
        }

        await createAnnouncement({
          courseId,
          payload: {
            announcementTitle,
            announcementContent,
          },
        }).unwrap();
        message.success(`Announcement ${announcementTitle} has been created!`);
      } catch (err) {
        message.error(getErrorMessage(err));
        throw err;
      }
    },
    [createAnnouncement, courseId],
  );

  return (
    <MultistepFormModal
      title="Create Announcement"
      form={form}
      onSubmit={onFinish}
      buttonChildren="New"
      formSteps={[
        <>
          <StringFormItem label="Announcement Title" name="announcementTitle" isRequired />
          <StringFormItem label="Announcement Content" name="announcementContent" isRequired isTextArea />
        </>,
      ]}
    />
  );
}
