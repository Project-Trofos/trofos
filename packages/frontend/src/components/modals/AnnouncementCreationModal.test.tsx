import React from 'react';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';

import store from '../../app/store';
import server from '../../mocks/server';

import AnnouncementCreationModal from './AnnouncementCreationModal';

describe('test announcement creation modal', () => {
  // Establish API mocking before all tests.
  beforeAll(() => server.listen());

  // Reset any request handlers that we may add during the tests,
  // so they don't affect other tests.

  afterEach(() => server.resetHandlers());

  // Clean up after the tests are finished.
  afterAll(() => server.close());

  const setup = () => {
    const { baseElement, debug } = render(
      <Provider store={store}>
        <AnnouncementCreationModal courseId="1" />
      </Provider>,
    );
    return { baseElement, debug };
  };

  // Antd modal is rendered outside of `root` div and does not get unmounted after closing
  const expectModalInvisible = async (baseElement: HTMLElement) => {
    // eslint-disable-next-line testing-library/no-container, testing-library/no-node-access
    const modalWrapper = baseElement.getElementsByClassName('ant-modal-wrap');
    expect(modalWrapper.length).toBe(1);
    await waitFor(() => expect(modalWrapper[0]).toHaveStyle('display: none'));
  };

  it('should render modal with correct fields', () => {
    const { baseElement } = setup();

    // Open modal
    const button = screen.getByText(/new/i);
    fireEvent.click(button);

    // Ensure fields are present
    expect(screen.getByLabelText(/announcement title/i)).toBeInTheDocument();
    expect(screen.getByLabelText(/announcement content/i)).toBeInTheDocument();

    // Compare with snapshot to ensure structure remains the same
    expect(baseElement).toMatchSnapshot();
  });

  it('should require announcement title', async () => {
    setup();

    const button = screen.getByText(/new/i);
    fireEvent.click(button);

    const finishButton = await screen.findByText(/finish/i);
    fireEvent.click(finishButton);

    await screen.findByText('Please input announcement title!');
  });

  it('should require announcement content', async () => {
    setup();

    const button = screen.getByText(/new/i);
    fireEvent.click(button);

    const title = screen.getByLabelText(/announcement title/i);
    fireEvent.change(title, { target: { value: 'title' } });

    const finishButton = screen.getByText(/finish/i);
    fireEvent.click(finishButton);

    await screen.findByText('Please input announcement content!');
  });

  it('should submit correctly if fields are typed in', async () => {
    const { baseElement } = setup();

    const button = screen.getByText(/new/i);
    fireEvent.click(button);

    const finishButton = await screen.findByText(/finish/i);

    const title = screen.getByLabelText(/announcement title/i);
    fireEvent.change(title, { target: { value: 'title' } });

    const content = screen.getByLabelText(/announcement content/i);
    fireEvent.change(content, { target: { value: 'content' } });

    fireEvent.click(finishButton);

    // Modal is closed
    await expectModalInvisible(baseElement);
  });
});
