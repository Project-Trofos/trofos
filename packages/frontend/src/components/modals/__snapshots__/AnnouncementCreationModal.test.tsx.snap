// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test announcement creation modal > should render modal with correct fields 1`] = `
<body>
  <div>
    <button
      aria-label="open-form"
      class="ant-btn ant-btn-default"
      type="button"
    >
      <span>
        New
      </span>
    </button>
  </div>
  <div>
    <div
      class="ant-modal-root"
    >
      <div
        class="ant-modal-mask"
      />
      <div
        class="ant-modal-wrap"
        tabindex="-1"
      >
        <div
          aria-labelledby="test-id"
          aria-modal="true"
          class="ant-modal"
          role="dialog"
          style="width: 520px;"
        >
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
          <div
            class="ant-modal-content"
          >
            <button
              aria-label="Close"
              class="ant-modal-close"
              type="button"
            >
              <span
                class="ant-modal-close-x"
              >
                <span
                  aria-label="close"
                  class="anticon anticon-close ant-modal-close-icon"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="close"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                    />
                  </svg>
                </span>
              </span>
            </button>
            <div
              class="ant-modal-header"
            >
              <div
                class="ant-modal-title"
                id="test-id"
              >
                Create Announcement
              </div>
            </div>
            <div
              class="ant-modal-body"
            >
              <form
                autocomplete="off"
                class="ant-form ant-form-vertical"
                id="multi-step-modal-form"
              >
                <div
                  class="ant-form-item"
                >
                  <div
                    class="ant-row ant-form-item-row"
                  >
                    <div
                      class="ant-col ant-form-item-label"
                    >
                      <label
                        class="ant-form-item-required"
                        for="multi-step-modal-form_announcementTitle"
                        title="Announcement Title"
                      >
                        Announcement Title
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control"
                    >
                      <div
                        class="ant-form-item-control-input"
                      >
                        <div
                          class="ant-form-item-control-input-content"
                        >
                          <input
                            aria-required="true"
                            class="ant-input"
                            id="multi-step-modal-form_announcementTitle"
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-form-item"
                >
                  <div
                    class="ant-row ant-form-item-row"
                  >
                    <div
                      class="ant-col ant-form-item-label"
                    >
                      <label
                        class="ant-form-item-required"
                        for="multi-step-modal-form_announcementContent"
                        title="Announcement Content"
                      >
                        Announcement Content
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control"
                    >
                      <div
                        class="ant-form-item-control-input"
                      >
                        <div
                          class="ant-form-item-control-input-content"
                        >
                          <textarea
                            aria-required="true"
                            class="ant-input"
                            id="multi-step-modal-form_announcementContent"
                            rows="4"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="ant-modal-footer"
            >
              <button
                class="ant-btn ant-btn-primary"
                form="multi-step-modal-form"
                type="button"
              >
                <span>
                  Finish
                </span>
              </button>
            </div>
          </div>
          <div
            aria-hidden="true"
            style="width: 0px; height: 0px; overflow: hidden; outline: none;"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
